### ISSUE TITLE
Linkerd initial memory usage higher in latest versions

### ISSUE THREAD

**Issue by DukeyToo** [linkerd/linkerd#2180]
Issue Type:

- [x] Bug report

**What happened**:

We are trying to upgrade from linkerd 1.4.1 to 1.5.2. In doing so, we noticed the Java RSS memory usage was larger on the new instances, even though we have not changed any memory settings or linkerd yaml.

**What you expected to happen**:

We expected memory usage to be consistent between linkerd versions, with possible minor differences.

**How to reproduce it (as minimally and precisely as possible)**:

Environment: 
```
JVM_HEAP_MAX=1024M
JVM_HEAP_MIN=1024M
FINAGLE_WORKERS=24
LOCAL_JVM_OPTIONS=-XX:NativeMemoryTracking=summary
```

Startup linkerd on a box with JDK 8 (I'm using openjdk), and note the process-id. No need to run any traffic through it. Run command `jcmd <your-pid> VM.native_memory summary` and note the total memory committed to linkerd. In 1.5.2, this is what I see:

```
Native Memory Tracking:

Total: reserved=3166475KB, committed=2079147KB
-                 Java Heap (reserved=1048576KB, committed=1048576KB)
                            (mmap: reserved=1048576KB, committed=1048576KB)

-                     Class (reserved=1101695KB, committed=61231KB)
                            (classes #9424)
                            (malloc=1919KB #13739)
                            (mmap: reserved=1099776KB, committed=59312KB)

-                    Thread (reserved=83571KB, committed=83571KB)
                            (thread #82)
                            (stack: reserved=83204KB, committed=83204KB)
                            (malloc=272KB #418)
                            (arena=95KB #160)

-                      Code (reserved=50449KB, committed=3585KB)
                            (malloc=529KB #2000)
                            (mmap: reserved=49920KB, committed=3056KB)

-                        GC (reserved=21756KB, committed=21756KB)
                            (malloc=18336KB #308)
                            (mmap: reserved=3420KB, committed=3420KB)

-                  Compiler (reserved=249KB, committed=249KB)
                            (malloc=119KB #379)
                            (arena=130KB #4)

-                  Internal (reserved=800308KB, committed=800308KB)
                            (malloc=800308KB #63799)

-                    Symbol (reserved=13036KB, committed=13036KB)
                            (malloc=11621KB #90207)
                            (arena=1415KB #1)

-    Native Memory Tracking (reserved=2688KB, committed=2688KB)
                            (malloc=13KB #147)
                            (tracking overhead=2675KB)

-               Arena Chunk (reserved=253KB, committed=253KB)
                            (malloc=253KB)

-                   Unknown (reserved=43896KB, committed=43896KB)
                            (mmap: reserved=43896KB, committed=43896KB)
```

Of note, "Internal" has reserved 800MB on top of the 1000MB heap. I don't have an equivalent output from linkerd 1.4.1, but I know the RSS total was very close to the heap size (not double the heap size).

**Environment**

I'll work on a minimal reproducible config example. For now, I'll say that we're running linkerd, connecting to namerd, which resolves mesos marathon apps.

**Comment by ganasubrgit** [comment]
jcmd output from my docker linkerd

```
root@ce01267112ae:/io.buoyant/linkerd/1.5.0# jcmd 1 VM.native_memory summary
1:

Native Memory Tracking:

Total: reserved=2745121KB, committed=1655117KB
-                 Java Heap (reserved=1048576KB, committed=1048576KB)
                            (mmap: reserved=1048576KB, committed=1048576KB)

-                     Class (reserved=1088396KB, committed=45776KB)
                            (classes #7364)
                            (malloc=908KB #7016)
                            (mmap: reserved=1087488KB, committed=44868KB)

-                    Thread (reserved=24777KB, committed=24777KB)
                            (thread #25)
                            (stack: reserved=24672KB, committed=24672KB)
                            (malloc=76KB #127)
                            (arena=28KB #48)

-                      Code (reserved=50156KB, committed=2772KB)
                            (malloc=236KB #1162)
                            (mmap: reserved=49920KB, committed=2536KB)

-                        GC (reserved=12840KB, committed=12840KB)
                            (malloc=9404KB #207)
                            (mmap: reserved=3436KB, committed=3436KB)

-                  Compiler (reserved=163KB, committed=163KB)
                            (malloc=33KB #116)
                            (arena=130KB #2)

-                  Internal (reserved=463267KB, committed=463267KB)
                            (malloc=463267KB #40876)

-                    Symbol (reserved=10767KB, committed=10767KB)
                            (malloc=9544KB #74950)
                            (arena=1223KB #1)

-    Native Memory Tracking (reserved=1954KB, committed=1954KB)
                            (malloc=6KB #78)
                            (tracking overhead=1947KB)

-               Arena Chunk (reserved=186KB, committed=186KB)
                            (malloc=186KB)

-                   Unknown (reserved=44040KB, committed=44040KB)
                            (mmap: reserved=44040KB, committed=44040KB)
```

**Comment by ganasubrgit** [comment]
git repo => https://git.target.com/GanapathySanthanam/docker-linkerd

**Comment by DukeyToo** [comment]
Thanks @ganasubrgit for help reproducing - the above "docker linkerd" minimal example is from the following linkerd yaml: 

```
admin:
  ip: 0.0.0.0
  port: 9990

routers:
- protocol: http
  interpreter:
    kind: io.l5d.namerd.http
    experimental: true
    dst: /$/inet/n5d/4180
  servers:
  - port: 4140
    ip: 0.0.0.0

usage:
  enabled: false
```

The "internal" committed memory on that is about 450MB, which is not as bad as my first example, but still illustrates the issue of linkerd reserving a lot of memory on top of what we assigned for heap.

**Comment by DukeyToo** [comment]
Full reproducible example is here: https://github.com/DukeyToo/linkerd-docker-example - thanks again @ganasubrgit for putting that together.

To reproduce, change linkerd version as needed in the docker compose file, then run:

```
docker-compose up -d
docker exec -it docker-linkerd_l5d_1 /bin/bash
apt-get update && apt-get install -y openjdk-8-jdk
jcmd 1 VM.native_memory summary
```

**Comment by dadjeibaah** [comment]
Managed to track this down and found out that this memory increase is caused by this netty [issue](https://github.com/netty/netty/issues/8159) and has been fixed since Netty 4.1.31. The finagle repo also has an issue that reported a similar problem [here](https://github.com/twitter/finagle/issues/738) and they also have recently picked up the Netty fix.

It looks like the most recent version of Finagle (18.12.0) corrects this increased non-heap memory and upgrading Linkerd to use Finagle 18.12.0 will ultimately fix this issue. Yay to library dependencies 😅!

**Comment by dadjeibaah** [comment]
Can confirm that the finagle upgrade fixes this issue. Running a docker image of this PR #2194 yields this native memory summary:
```bash
Printing VM memory summary...
1:

Native Memory Tracking:

Total: reserved=2290251KB, committed=1200727KB
-                 Java Heap (reserved=1048576KB, committed=1048576KB)
                            (mmap: reserved=1048576KB, committed=1048576KB)

-                     Class (reserved=1092594KB, committed=50454KB)
                            (classes #8180)
                            (malloc=1010KB #7916)
                            (mmap: reserved=1091584KB, committed=49444KB)

-                    Thread (reserved=24777KB, committed=24777KB)
                            (thread #25)
                            (stack: reserved=24672KB, committed=24672KB)
                            (malloc=77KB #125)
                            (arena=28KB #48)

-                      Code (reserved=50186KB, committed=2802KB)
                            (malloc=266KB #1127)
                            (mmap: reserved=49920KB, committed=2536KB)

-                        GC (reserved=8398KB, committed=8398KB)
                            (malloc=4630KB #212)
                            (mmap: reserved=3768KB, committed=3768KB)

-                  Compiler (reserved=169KB, committed=169KB)
                            (malloc=39KB #105)
                            (arena=130KB #2)

-                  Internal (reserved=5283KB, committed=5283KB)
                            (malloc=5283KB #13480)

-                    Symbol (reserved=11712KB, committed=11712KB)
                            (malloc=10266KB #80804)
                            (arena=1447KB #1)

-    Native Memory Tracking (reserved=1631KB, committed=1631KB)
                            (malloc=7KB #82)
                            (tracking overhead=1624KB)

-               Arena Chunk (reserved=185KB, committed=185KB)
                            (malloc=185KB)

-                   Unknown (reserved=46740KB, committed=46740KB)
                            (mmap: reserved=46740KB, committed=46740KB)
```

**Comment by adleong** [comment]
Will be fixed by #2194
